File: 그래프_bfs, 최단거리문제/leetcode1129(Medium)-Shortest Path with Alternating Colors.js
// 119ms(33.33%), 50.8MB(11.49%)
class MyNode {
  constructor(value) {
    this.value = value
    this.next = null
  }
}

class MyQueue {
  constructor() {
    this.front = this.tail = null
    this.size = 0
  }

  enqueue(value) {
    const node = new MyNode(value)
    if (!this.front) {
      this.front = this.tail = node
    } else {
      this.tail = this.tail.next = node
    }
    this.size += 1
  }

  dequeue() {
    if (!this.front) {
      return undefined
    }
    const result = this.front.value
    this.front = this.front.next
    this.size -= 1
    return result
  }
}

const initiailizeGraph = (n) => {
  const graph = {}
  for (let i = 0; i < n; i += 1) {
    graph[i] = []
  }
  return graph
}

/**
 * @param {number} n
 * @param {number[][]} redEdges
 * @param {number[][]} blueEdges
 * @return {number[]}
 */
var shortestAlternatingPaths = function (n, redEdges, blueEdges) {
  const RED = 0
  const BLUE = 1
  const graph = initiailizeGraph(n)
  for (const [v, w] of redEdges) {
    graph[v].push([w, RED])
  }
  for (const [v, w] of blueEdges) {
    graph[v].push([w, BLUE])
  }

  const visited = Array.from({ length: n }, () => new Array(2).fill(false))
  const answer = new Array(n).fill(Infinity)
  answer[0] = 0
  const q = new MyQueue()
  for (const [_, color] of graph[0]) {
    visited[0][color] = true
    q.enqueue([0, null, 0]) // node, unknown-color, dist
  }

  while (q.size > 0) {
    const [v, prevColor, dist] = q.dequeue()
    for (const [w, color] of graph[v]) {
      if (color !== prevColor && !visited[w][color]) {
        visited[w][color] = true
        q.enqueue([w, color, dist + 1])
        answer[w] = Math.min(answer[w], dist + 1)
      }
    }
  }

  return answer.map((x) => (x === Infinity ? -1 : x))
}

File: 그래프_bfs, 최단거리문제/leetcode1129(Medium)-Shortest Path with Alternating Colors.py
# 88ms(77.86%), 14.2MB(73.17%)
class Solution:
    def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
        INF = int(1e9)
        RED = 0
        BLUE = 1
        graph = collections.defaultdict(list)
        for v, w in redEdges:
            graph[v].append([w, RED])
        for v, w in blueEdges:
            graph[v].append([w, BLUE])

        visited = [[False, False] for _ in range(n)]
        answer = [INF] * n
        answer[0] = 0
        q = collections.deque()
        # start to 0 Node
        for w, color in graph[0]:
            visited[0][color] = True
            q.append([0, None, 0]) # node, unknown-color, dist
        
        # BFS
        while q:
            v, prev_color, dist = q.popleft()
            for w, color in graph[v]:
                if color != prev_color and not visited[w][color]:
                    visited[w][color] = True
                    q.append([w, color, dist + 1])
                    answer[w] = min(answer[w], dist + 1)
        
        # if element is INF, not path.
        return [-1 if x == INF else x for x in answer]